Name: Neha Kale
Objective: Separating Spam From Ham

### Task 1.1 - Loading the Dataset

In [ ]:
#https://ocw.mit.edu/courses/sloan-school-of-management/15-071-the-analytics-edge-spring-2017/text-analytics/assignment-5/

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
%matplotlib inline

In [ ]:
from google.colab import drive #mounting google drive which contains the dataset 
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
%cd /content/drive/MyDrive/DS

/content/drive/MyDrive/DS


In [ ]:
%pwd # checking the present directory which contains the dataset

'/content/drive/MyDrive/DS'

In [ ]:
df = pd.read_csv('emails.csv',encoding='cp1252', delimiter=',', quotechar='"') 

In [ ]:
df

text  spam
0     Subject: naturally irresistible your corporate...     1
1     Subject: the stock trading gunslinger  fanny i...     1
2     Subject: unbelievable new homes made easy  im ...     1
3     Subject: 4 color printing special  request add...     1
4     Subject: do not have money , get software cds ...     1
...                                                 ...   ...
5723  Subject: re : research and development charges...     0
5724  Subject: re : receipts from visit  jim ,  than...     0
5725  Subject: re : enron case study update  wow ! a...     0
5726  Subject: re : interest  david ,  please , call...     0
5727  Subject: news : aurora 5 . 2 update  aurora ve...     0

[5728 rows x 2 columns]

In [ ]:
print("Shape of data (samples, features): ",df.shape)

Shape of data (samples, features):  (5728, 2)


Question 1: How many emails are in the dataset? 
- Thus we can see there are 5728 emails in the given dataset 

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5728 entries, 0 to 5727
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   text    5728 non-null   object
 1   spam    5728 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 89.6+ KB


In [ ]:
df['spam'].value_counts()

0    4360
1    1368
Name: spam, dtype: int64

Question 2:How many of the emails are spam?  
- 1368 emails are spam 

In [ ]:
first_w = df.apply(lambda x:x['text'].split()[0],axis=1)
first_w.value_counts()

Subject:    5728
dtype: int64

In [ ]:
# %matplotlib inline 
# sns.set(rc = {'figure.figsize':(11.7,8.27)})
# sns.set_style('darkgrid')
# nlp_words = nltk.FreqDist(first_words)
# nlp_words.plot(20) 

In [ ]:
max_characters = 0 
for i,email in enumerate(df['text']):
  max_characters = max(max_characters,len(email))

max_characters

43952

###Task 2.1 - Preparing the Corpus 

In [ ]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [ ]:
import string
from nltk.corpus import stopwords
from nltk import PorterStemmer as Stemmer

def preprocess(text):
    # Converting to lowercase
    text = text.lower()
    # Removing punctuation
    text = ''.join([t for t in text if t not in string.punctuation])
    # Removing stopwords
    text = [t for t in text.split() if t not in stopwords.words('english')]
    ##Applying stemming
    st = Stemmer()
    text = [st.stem(t) for t in text]
    text = ' '.join(text)
    return text

In [ ]:
corpus = df.apply(lambda x:preprocess(x['text']),axis = 1)

In [ ]:
corpus

0       subject natur irresist corpor ident lt realli ...
1       subject stock trade gunsling fanni merril muzo...
2       subject unbeliev new home made easi im want sh...
3       subject 4 color print special request addit in...
4       subject money get softwar cd softwar compat gr...
                              ...                        
5723    subject research develop charg gpg forward shi...
5724    subject receipt visit jim thank invit visit ls...
5725    subject enron case studi updat wow day super t...
5726    subject interest david pleas call shirley cren...
5727    subject news aurora 5 2 updat aurora version 5...
Length: 5728, dtype: object

In [ ]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer 

def fn_tdm_df(docs):
    ''' create a term document matrix as pandas DataFrame '''
    #initialize the  vectorizer
    vectorizer = CountVectorizer()
    x1 = vectorizer.fit_transform(docs)
    #create dataFrame
    df2 = pd.DataFrame(x1.toarray().transpose(), index = vectorizer.get_feature_names())
    return df2,x1

In [ ]:
dtm,X_transform = fn_tdm_df(corpus)

In [ ]:
dtm.to_csv('dtm.csv')

In [ ]:
dtm

0     1     2     3     4     5     6     7     8     9     ...  \
00           0     0     0     0     0     0     0     0     0     0  ...   
000          0     0     0     0     0     0     0     0     1     0  ...   
0000         0     0     0     0     0     0     0     0     0     0  ...   
000000       0     0     0     0     0     0     0     0     0     0  ...   
00000000     0     0     0     0     0     0     0     0     0     0  ...   
...        ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   
zymg         0     0     0     0     0     0     0     0     0     0  ...   
zzmacmac     0     0     0     0     0     0     0     0     0     0  ...   
zzn          0     0     0     0     0     0     0     0     0     0  ...   
zzncacst     0     0     0     0     0     0     0     0     0     0  ...   
zzzz         0     0     0     0     0     0     0     0     0     0  ...   

          5718  5719  5720  5721  5722  5723  5724  5725  5726  5727  
00           0     0     0     0     0     0     1     4     0     0  
000          1     0     0     0     0     0     0     0     0     0  
0000         0     0     0     0     0     0     0     0     0     0  
000000       0     0     0     0     0     0     0     0     0     0  
00000000     0     0     0     0     0     0     0     0     0     0  
...        ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
zymg         0     0     0     0     0     0     0     0     0     0  
zzmacmac     0     0     0     0     0     0     0     0     0     0  
zzn          0     0     0     0     0     0     0     0     0     0  
zzncacst     0     0     0     0     0     0     0     0     0     0  
zzzz         0     0     0     0     0     0     0     0     0     0  

[29254 rows x 5728 columns]

In [ ]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer 

def sparse_tdm_df(docs):
    ''' create a term document matrix as pandas DataFrame '''
    #initialize the  vectorizer
    vectorizer = CountVectorizer(min_df = 0.05)
    x1 = vectorizer.fit_transform(docs)
    #create dataFrame
    df2 = pd.DataFrame(x1.toarray().transpose(), index = vectorizer.get_feature_names())
    return df2

In [ ]:
sparse_dtm = sparse_tdm_df(corpus)
sparse_dtm

0     1     2     3     4     5     6     7     8     9     ...  5718  \
00        0     0     0     0     0     0     0     0     0     0  ...     0   
000       0     0     0     0     0     0     0     0     1     0  ...     1   
01        0     0     0     0     0     0     0     0     0     0  ...     0   
02        0     0     0     0     0     0     0     0     0     0  ...     0   
03        0     0     0     0     0     0     0     0     0     0  ...     0   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...   
work      0     0     0     0     0     0     1     0     0     0  ...     2   
would     0     0     0     0     0     0     0     0     0     0  ...     0   
write     0     0     0     0     0     0     0     0     0     0  ...     0   
www       0     0     0     0     0     0     0     0     0     0  ...     0   
year      0     0     0     0     0     0     0     0     0     0  ...     1   

       5719  5720  5721  5722  5723  5724  5725  5726  5727  
00        0     0     0     0     0     1     4     0     0  
000       0     0     0     0     0     0     0     0     0  
01        0     2     0     0     0     0     1     0     0  
02        1     2     0     0     1     1     0     0     0  
03        6     0     0     3     3     0     0     0     0  
...     ...   ...   ...   ...   ...   ...   ...   ...   ...  
work      0     1     0     0     0     0     1     0     0  
would     1     1     0     1     1     0     0     0     0  
write     0     0     0     0     0     0     1     0     0  
www       0     0     0     0     0     0     0     0     2  
year      0     0     0     0     0     0     0     0     1  

[366 rows x 5728 columns]

Task 3.1 Building machine learning models 

In [ ]:
from sklearn.feature_extraction.text import TfidfTransformer
tfidf_transformer = TfidfTransformer().fit(X_transform)
X_tfidf = tfidf_transformer.transform(X_transform)
print(X_tfidf.shape)

(5728, 29254)


In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_transform, df['spam'], test_size=0.30, random_state=123)

In [ ]:
from sklearn import tree

spamCART = tree.DecisionTreeClassifier() # threshold of 0.5 for predictions is set by default
spamCART = spamCART.fit(X_train, y_train)
pred = spamCART.predict(X_test)

In [ ]:
import matplotlib.pyplot as plt
plt.figure(figsize = (100,100))
tree.plot_tree(spamCART)
plt.show()

In [ ]:
from sklearn.metrics import classification_report # for model evaluation metrics

print('*************** Evaluation on Test Data ***************')
score_te = spamCART.score(X_test, y_test)
print('Accuracy Score: ', score_te)
# Look at classification report to evaluate the model
print(classification_report(y_test, pred))
print('--------------------------------------------------------')
print("")

*************** Evaluation on Test Data ***************
Accuracy Score:  0.9616055846422339
              precision    recall  f1-score   support

           0       0.97      0.98      0.98      1319
           1       0.92      0.91      0.92       400

    accuracy                           0.96      1719
   macro avg       0.95      0.94      0.95      1719
weighted avg       0.96      0.96      0.96      1719

--------------------------------------------------------



In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_transform, df['spam'], test_size=0.30, random_state=123)

In [ ]:
#Import Random Forest Model
from sklearn.ensemble import RandomForestClassifier
#Create a Gaussian Classifier
spamRF =RandomForestClassifier(n_estimators=100)
#Train the model using the training sets y_pred=clf.predict(X_test)
spamRF.fit(X_train,y_train)
y_pred=spamRF.predict(X_test)

In [ ]:
#Import scikit-learn metrics module for accuracy calculation
from sklearn import metrics
# Model Accuracy, how often is the classifier correct?
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.9784758580570099


What is the training set accuracy of spamCART, using a threshold of 0.5 for predictions? 
96.10% 

What is the training set accuracy of spamRF, using a threshold of 0.5 for predictions?
97.84% 


In [ ]:

fpr = dict()
tpr = dict()
roc_auc = dict()
for i in range(2):
    fpr[i], tpr[i], _ = roc_curve(y_test[:, i], y_score[:, i])
    roc_auc[i] = auc(fpr[i], tpr[i])

# Compute micro-average ROC curve and ROC area
fpr["micro"], tpr["micro"], _ = roc_curve(y_test.ravel(), y_score.ravel())
roc_auc["micro"] = auc(fpr["micro"], tpr["micro"])

#ROC curve for a specific class here for the class 2
roc_auc[2]

NameError: ignored